[SPARK-51423][SQL] Add the current_time() function for TIME datatype

the-sakthi opened the pull request:
### What changes were proposed in this pull request?
This PR adds support for a new function current_time() which returns the current time at the start of query evaluation.

```bash
# happy cases
scala> spark.sql("SELECT current_time(0);").show()
+---------------+
|current_time(0)|
+---------------+
|       17:11:26|
+---------------+

scala> spark.sql("SELECT current_time(3);").show()
+---------------+
|current_time(3)|
+---------------+
|   17:11:50.225|
+---------------+

scala> spark.sql("SELECT current_time(6);").show()
+---------------+
|current_time(6)|
+---------------+
|17:12:00.734735|
+---------------+

# No braces and Empty braces
scala> spark.sql("SELECT current_time;").show()
+---------------+
|current_time(6)|
+---------------+
|17:12:23.132088|
+---------------+


scala> spark.sql("SELECT current_time();").show()
+---------------+
|current_time(6)|
+---------------+
|17:12:26.718602|
+---------------+

# foldability
## Nested Arithmetic
scala> spark.sql("SELECT current_time((4 - 2) * (1 + 1));").show()
+---------------------------------+
|current_time(((4 - 2) * (1 + 1)))|
+---------------------------------+
|                    17:13:04.4647|
+---------------------------------+

## Casting String literals
scala> spark.sql("SELECT current_time(CAST(' 0005 ' AS INT));").show()
+---------------------------------+
|current_time(CAST( 0005  AS INT))|
+---------------------------------+
|                   17:13:26.28039|
+---------------------------------+

scala> spark.sql("SELECT current_time('5');").show()
+---------------+
|current_time(5)|
+---------------+
| 22:34:07.65007|
+---------------+


## Combine Cast and Arithmetic
scala> spark.sql("SELECT current_time(CAST('4' AS INT) * CAST('1' AS INT));").show()
+-----------------------------------------------+
|current_time((CAST(4 AS INT) * CAST(1 AS INT)))|
+-----------------------------------------------+
|                                  17:14:06.7151|
+-----------------------------------------------+

# failure cases
scala> spark.sql("SELECT current_time(-1);").show()
org.apache.spark.sql.catalyst.ExtendedAnalysisException: [DATATYPE_MISMATCH.VALUE_OUT_OF_RANGE] Cannot resolve "current_time(-1)" due to data type mismatch: The `precision` must be between [0, 6] (current value = -1). SQLSTATE: 42K09; line 1 pos 7;
'Project [unresolvedalias(current_time(-1))]
+- OneRowRelation'

scala> spark.sql("SELECT current_time('foo');").show()
org.apache.spark.SparkNumberFormatException: [CAST_INVALID_INPUT] The value 'foo' of the type "STRING" cannot be cast to "INT" because it is malformed. Correct the value as per the syntax, or change its target type. Use `try_cast` to tolerate malformed input and return NULL instead. SQLSTATE: 22018
== SQL (line 1, position 8) ==
SELECT current_time('foo');

scala> spark.sql("SELECT current_time(2,2);").show()
org.apache.spark.sql.AnalysisException: [WRONG_NUM_ARGS.WITHOUT_SUGGESTION] The `current_time` requires [0, 1] parameters but the actual number is 2. Please, refer to 'https://spark.apache.org/docs/latest/sql-ref-functions.html' for a fix. SQLSTATE: 42605; line 1 pos 7


# All calls of current_time within the same query should return the same value.
scala> val df = spark.sql("""
     |   SELECT
     |     current_time AS col1,
     |     current_time() AS col2,
     |  current_time(0) AS col3,
     |  current_time(1) AS col4,
     |  current_time(2) AS col5,
     |  current_time(3) AS col6,
     |  current_time(4) AS col7,
     |  current_time(5) AS col8,
     |  current_time(6) AS col9,
     |     current_time AS col10
     | """)
val df: org.apache.spark.sql.DataFrame = [col1: time(6), col2: time(6) ... 8 more fields]

scala> df.show()
+---------------+---------------+--------+----------+-----------+-----------+-------------+--------------+---------------+---------------+
|           col1|           col2|    col3|      col4|       col5|       col6|         col7|          col8|           col9|          col10|
+---------------+---------------+--------+----------+-----------+-----------+-------------+--------------+---------------+---------------+
|17:15:47.680648|17:15:47.680648|17:15:47|17:15:47.6|17:15:47.68|17:15:47.68|17:15:47.6806|17:15:47.68064|17:15:47.680648|17:15:47.680648|
+---------------+---------------+--------+----------+-----------+-----------+-------------+--------------+---------------+---------------+

```

### Why are the changes needed?
Adds a built-in current_time([n]) function returning just the time portion (in a TIME(n) type). This aligns Spark with other SQL systems offering a native time function, improves convenience for time-only queries, and complements existing functions like current_date and current_timestamp.


### Does this PR introduce _any_ user-facing change?
Yes, adds a new function. Users can now get the current time using this function.

### How was this patch tested?
Manual testing as shown above and running UTs added:

```bash
$ build/sbt "test:testOnly *TimeExpressionsSuite.scala"
$ build/sbt "test:testOnly *ComputeCurrentTimeSuite.scala"
$ build/sbt "test:testOnly *ResolveInlineTablesSuite.scala
$ build/sbt "test:testOnly *AnalysisSuite.scala
```


### Was this patch authored or co-authored using generative AI tooling?
No

MaxGekk reviewed (changes requested):
> ... which returns the current time at the start of query evaluation.

I think the current implementation doesn't satisfy the requirement, and every call in a query returns new current time. Could you look at the `CurrentTimestampLike`, and the rule:
https://github.com/apache/spark/blob/2822526d6651b129943aec130860106bc139ec42/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala#L108-L111

the-sakthi commented:
Ahh sure! Thanks for the early review, Max. Will refer `CurrentTimestampLike` to better cater to the requirement.

the-sakthi committed:
[SPARK-51423][SQL] Add the current_time() function for TIME datatype

the-sakthi commented:
Thanks for the feedback @MaxGekk , let me know how this one looks!

MaxGekk commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:
```diff
@@ -133,6 +134,32 @@ trait SparkDateTimeUtils {
+   * becomes `123000` microseconds (12:34:56.123).
+   *
+   * @param micros The original time in microseconds.
+   * @param p      The fractional second precision (range 0 to 6).
```
Could you add an assert for that.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/timeExpressions.scala:
```diff
@@ -290,3 +291,63 @@ object HourExpressionBuilder extends ExpressionBuilder {
+  group = "datetime_funcs",
+  since = "4.1.0")
+// scalastyle:on line.size.limit
+object CurrentTimeExpressionBuilder extends ExpressionBuilder {
```
You can just add constructors with different parameters to `CurrentTime` instead of using `ExpressionBuilder`.

the-sakthi replied:
Makes sense! Rev 2 accordingly.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:
```diff
@@ -129,6 +131,11 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
                 DateTimeUtils.microsToDays(currentTimestampMicros, cd.zoneId), DateType)
             })
+          case currentTimeType : CurrentTime =>
+            // scalastyle:off line.size.limit
```
Just put split the line below, and remove this.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:
```diff
@@ -129,6 +131,11 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
             })
+          case currentTimeType : CurrentTime =>
+            // scalastyle:off line.size.limit
+            val truncatedTime = truncateTimeMicrosToPrecision(currentTimeOfDayMicros, currentTimeType.precision)
```
```suggestion
            val truncatedTime = truncateTimeMicrosToPrecision(currentTimeOfDayMicros,
              currentTimeType.precision)```

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -2148,6 +2148,21 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
   }
 
+  // scalastyle:off line.size.limit
+  def nonFoldableInputError(argumentName: String, expression: Expression, requiredType: DataType): Throwable = {
```
Just put parameters on separate lines, see https://github.com/databricks/scala-style-guide?tab=readme-ov-file#spacing-and-indentation

the-sakthi replied:
Very helpful resource. Thanks, Max.

the-sakthi commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:
```diff
@@ -133,6 +134,32 @@ trait SparkDateTimeUtils {
+   * Gets the number of microseconds since midnight using the session time zone.
+   */
+  def instantToMicrosOfDay(instant: Instant): Long = {
+    val zoneId = getZoneId(SqlApiConf.get.sessionLocalTimeZone)
```
Ah sure. Makes sense.

the-sakthi committed:
Revision comments 1

- Handling foldable expressions via multiple constructors approach
  rather than expressionBuilder
- Added more tests
- Split lines to avoid using scalastyle line limit config
- Other review comments

the-sakthi commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/timeExpressions.scala:
```diff
@@ -290,3 +291,63 @@ object HourExpressionBuilder extends ExpressionBuilder {
+
+      case Seq(_) =>
+        val child = expressions.head
+        throw QueryCompilationErrors.nonFoldableInputError("precision", child, IntegerType)
```
Ah totally! Have catered to this in the new revision.

the-sakthi commented:
Thanks for the feedback @MaxGekk , let me know how this revision one looks! Hopefully this is better aligned to the review comments!

the-sakthi committed:
Add CurrentTime to ignoreset of examples for ExpressionInfoSuite

the-sakthi committed:
modify ExpressionParserSuite and SQLQueryTestHelper

MaxGekk commented:
@the-sakthi Thank you for the ping. I am traveling right now. I will look at the PR slightly latter.

the-sakthi commented:
Thanks for the response, Max, even while traveling. Will await your review!

MaxGekk reviewed (commented):
lgtm in general. Let me review it again tomorrow.

the-sakthi commented:
Sure thanks @MaxGekk ! Will wait.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/timeExpressions.scala:
```diff
@@ -290,3 +294,100 @@ object HourExpressionBuilder extends ExpressionBuilder {
+
+@ExpressionDescription(
+  usage = """
+    _FUNC_([precision]) - Returns the current time at the start of query evaluation.
```
It would be nice if you describe the `precision` parameter since this is visible to users. Could you `arguments` like
https://github.com/apache/spark/blob/f4ce0ab8cee043d0b1a8d935fb13ce617d2de873/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/timeExpressions.scala#L136-L141

the-sakthi replied:
Definitely!

MaxGekk reviewed (commented):
LGTM except of a comment.

MaxGekk commented:
@the-sakthi Please, resolve conflicts.

the-sakthi committed:
Revision Comments 2: Add user facing argument description

the-sakthi commented:
@MaxGekk The latest push has the updated user description and rebased with latest main branch. Let me know!
Thanks.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/timeExpressions.scala:
```diff
@@ -349,3 +353,109 @@ object SecondExpressionBuilder extends ExpressionBuilder {
+  """,
+  arguments = """
+    Arguments:
+      * precision - An optional integer literal in the range [${TimeType.MIN_PRECISION}..${TimeType.MICROS_PRECISION}], indicating how many
```
I guess the constants ${TimeType.MIN_PRECISION} won't embedded in the string. It should start from s" at least. Could you generate the doc, and double check.

the-sakthi replied:
Changed to hardcoded versions in the new revision.

the-sakthi committed:
Hardcode docstring values

MaxGekk commented:
+1, LGTM. Merging to master.
Thank you, @the-sakthi.

the-sakthi commented:
Thank you very much @MaxGekk !
